### DISCUSSION TITLE
Duplicate joystick button press oddity

### DISCUSSION THREAD

**Discussion by mattcaron** [joncampbell123/dosbox-x#4594] [General]
This is admittedly somewhat vague and ill-formed, not yet to the level of a bug report.

Off the bat:

- Ubuntu 22.04
- v2023.10.06
- I tried this with both SDL1 and SDL2 builds

I have a relatively standard modern HOTAS setup that I am mapping via 2axis. Broadly speaking, it works, but I have a very odd thing happening and I wonder if people might have some insight.

On my throttle, it has ALL THE BUTTONS. Seriously, it's 5 axes, 79 buttons. (VKB STECS MAX)

When trying to map things, I'll map some button, say 15 to a keyboard press (we'll call the key K here). Works fine. Mapper sees it, mapper blinks the light when I press it, and then it works in game.

I then go to map more buttons, all works fine. But then I hit another one, say, 52. Now, 52 isn't bound to anything. But, if I press it, the key mapped to 15 lights up! Indeed, if I then bind it to something (we'll call that key L here), the mapper reports having bound 15!

If I have the mapper open in a window, and `jstest-gtk` open in another window, I'll press 15:

- in the mapper, both keys K & L illuminates, registering presses
- in `jstest-gtk`, button 15 illuminates, registering a press there.

But, if I press 52:

- in the mapper, both keys K & L illuminates, registering presses
- in `jstest-gtk`, button 52 illuminates, registering a press there.

So, `jstest-gtk` can distinguish between the two buttons, but DOSBox-X seemingly cannot.

There's also no rhyme or reason to it. My first thought was that it might be a wrapping issue - say, everything over X gets wrapped to N % X because the input button indices are limited to some limit (kind of like how `buttonwrap` works, but on the input, rather than output, side), but I've had there be "duplicate" buttons right in the middle of a range that otherwise maps just fine (of course, I haven't mapped all the buttons, only some).

Any thoughts? Even just a "look at this bit of code and change that and see if it makes a difference" would be appreciated.

Thanks in advance.

**Comment by rderooy** [comment]
To be honest, it does sound like some kind of wrapping issue. Joystick support is largely handled by SDL itself, although this may be a mapper issue.

Do you know if other SDL programs are able to handle the controller properly? For instance a recent development build of DOSBox Staging?

**Reply by mattcaron** [reply]
Yes. My standard test case for this is [sdl-jstest](https://github.com/Grumbel/sdl-jstest), which has both SDL1 and SDL2 versions. Testing both versions reports the same results as `jstest-gtk` - that is, correct behavior.

The fact that both SDL1 and SDL2 builds of dosbox-x has me thinking it's related to the mapper. I'm poking about `sdl_mapper.cpp`, see what I can see.

**Reply by mattcaron** [reply]
And, this looks promising:
```
       void ActivateJoystickBoundEvents() {
...
               for (i=0; i<button_cap; i++) {
            if (SDL_JoystickGetButton(sdl_joystick,(int)i))
                button_pressed[i % button_wrap]=true;
        }
```
And looking those up....
```
        buttons=(Bitu)SDL_JoystickNumButtons(sdl_joystick);
        button_wrap=buttons;
        button_cap=buttons;
        if (button_wrapping_enabled) {
            button_wrap=emulated_buttons;
            if (buttons>MAXBUTTON_CAP) button_cap = MAXBUTTON_CAP;
        }
        if (button_wrap > MAXBUTTON) button_wrap = MAXBUTTON;
```
1. Get number of buttons on joystick. Here, will be 79.
2. `button_wrap` and `button_cap` both set to that.
3. `button_wrapping_enabled` is disabled, skip that.
4. But that last one... what is `MAXBUTTON`? 32. 

I bet it's wrapping at 32 and I just wasn't clever enough to see it (likely because I was looking at 64).

I'm going to set it to 128 and do some testing - hopefully this weekend, but no promises. I have much to do around the house.

**Reply by mattcaron** [reply]
Bumping it to 128 fixes it. I can bind all the buttons. Jane's ATF will never be the same.

Since the code iterates over the whole array, I have a minor concern about increasing latency more than is strictly required. Why use 128 when 96 is a nice even multiple > 79?

I don't know if there are any setups with even more buttons, but this seems reasonable, and if one comes along we can bump it more.

```
-#define MAXBUTTON               32
+#define MAXBUTTON               96
```

I'll prepare a proper PR tomorrow, but I have another concern of it breaking users' current maps. As in - if a user has a stick with 35 buttons (0-34), buttons 0-31 will be mapped normally, but then it will wrap. Assuming they don't bind button 1, they can bind button 33 without issue as it will wrap to 1. This fix, however, will be a breaking change to such maps.

At the risk of creeping featurism, I wonder if we need to add a config variable for "max joystick input buttons" or similar. We hardcode MAXBUTTON to something sane (say, 128) because that gets our arrays all sized properly, and then the config value will determine just how much we iterate. If you're using a more normal setup (say, a CH stick and throttle, for that vintage 90's feel), then set to 32, which should cover all the buttons and keep the polling quite tight. But, if you're using something by VKB, Virpil, or similar, you can crank that up and get more buttons at the expense of polling.

Or am I overthinking it (probably because I'm used to writing firmware on things that run at 20Mhz and have 128KB RAM)?

**Comment by mattcaron** [comment]
I wanted to finish my episode of Beavis and Butthead (Black Sabbath was on..), so I wrote the PR.

https://github.com/joncampbell123/dosbox-x/pull/4596

**Reply by Torinde** [reply]
Great!
I'm wondering what's the maximum number of buttons joysticks have? What is the model with the most buttons? Are they always mapped to a keyboard key? If so, then 128 will be the absolute maximum?

Can you also please mark this discussion as closed/answered (since you solved the issue)?

**Reply by mattcaron** [reply]
1. The VKB Gladiator product line is modular, so functionally there is no limit.
2. I believe Windows has a 128 button limit. VKB's software allows splitting the controller into multiple virtual controllers to circumvent this limitation. So, setting it to 128 (to match Windows) might have made more sense (I went with 96.)

Will mark as resolved. Sorry for the delay. Forgot about it.
